clientele/framework/client.py

```diff
@@ -0,0 +1,406 @@
+from __future__ import annotations
+
+import inspect
+import re
+from contextlib import asynccontextmanager, contextmanager
+from dataclasses import dataclass
+from functools import wraps
+from typing import Any, Callable, Iterable, TypeVar, get_type_hints
+from urllib.parse import quote
+
+import httpx
+from pydantic import BaseModel
+
+from .config import Config
+
+try:  # pragma: no cover - conditional import
+    from pydantic import TypeAdapter
+
+    _HAS_TYPE_ADAPTER = True
+except ImportError:  # pragma: no cover - fallback for Pydantic v1
+    TypeAdapter = None  # type: ignore
+    _HAS_TYPE_ADAPTER = False
+
+try:  # pragma: no cover - conditional import
+    from pydantic.tools import parse_obj_as
+except Exception:  # pragma: no cover - fallback for Pydantic v2 only environments
+    parse_obj_as = None  # type: ignore
+
+
+_F = TypeVar("_F", bound=Callable[..., Any])
+_PATH_PARAM_PATTERN = re.compile(r"{([^{}]+)}")
+
+
+@dataclass
+class _RequestContext:
+    method: str
+    path_template: str
+    func: Callable[..., Any]
+    signature: inspect.Signature
+    type_hints: dict[str, Any]
+
+
+def _build_request_context(method: str, path: str, func: Callable[..., Any]) -> _RequestContext:
+    signature = inspect.signature(func)
+    type_hints = get_type_hints(func)
+    return _RequestContext(
+        method=method,
+        path_template=path,
+        func=func,
+        signature=signature,
+        type_hints=type_hints,
+    )
+
+
+@dataclass
+class _PreparedCall:
+    context: _RequestContext
+    bound_arguments: inspect.BoundArguments
+    call_arguments: dict[str, Any]
+    url_path: str
+    query_params: dict[str, Any] | None
+    data_payload: Any
+    headers_override: dict[str, str] | None
+    return_annotation: Any
+
+
+class Client:
+    """A decorator-driven HTTP client for sync and async requests.
+
+    Supports common HTTP verbs (GET, POST, PUT, PATCH, DELETE) and works with
+    both synchronous and ``async`` decorated callables without separate
+    classes. Decorated functions must have a real body; the wrapper executes
+    the HTTP request before calling the function and injects the parsed
+    response via a ``result`` argument when present.
+    """
+
+    def __init__(
+        self,
+        *,
+        config: Config | None = None,
+        base_url: str | None = None,
+        headers: dict[str, str] | None = None,
+        timeout: float | None = None,
+        follow_redirects: bool | None = None,
+        auth: httpx.Auth | tuple[str, str] | None = None,
+        verify: bool | str | None = None,
+        http2: bool | None = None,
+        limits: httpx.Limits | None = None,
+        proxies: httpx._types.ProxiesTypes | None = None,
+        transport: httpx.BaseTransport | httpx.AsyncBaseTransport | None = None,
+        cookies: httpx._types.CookieTypes | None = None,
+    ) -> None:
+        if config is None:
+            config = Config(
```

Comment:
I think what we should do is have a `DefaultConfig` somewhere in config that uses sensible defaults instead of setting it all here in the Client model, and then we can just import the default one.